File: packages/common/erp-common-ms-utils-js/src/helpers/env.ts
import dotenv from 'dotenv';
import { TypeErrorMessage } from '../errors/utils/error-type.enum';
import { sendError } from '../errors/utils/send-error';
import { IConfig } from '../interfaces/config.interface';
import { resolvePath, checkExistsFile } from './node';
import { initVault, Vault } from './vault';

export let env: NonNullable<IConfig>;

export const initEnv = async (useDotEnv = true) => {
  if (useDotEnv) {
    await readEnvFile();
  }
  const environment = process.env.ENVIRONMENT;
  if (!environment) {
    sendError(TypeErrorMessage.CONFIG);
  }

  const fileFolder =
    (environment == 'development' ? 'src' : 'dist') + '/environments/';
  const filePath = resolvePath(fileFolder + environment);
  const extension = environment == 'development' ? '.ts' : '.js';
  const fullFilePath = filePath + extension;

  if (!checkExistsFile(fullFilePath)) {
    sendError(TypeErrorMessage.CONFIG);
  }
  const { config } = await import(filePath);
  env = config;
  await enrichEnv(env, environment!);
};

const enrichEnv = async (env: any, environment: string) => {
  try {
    await initVault();
    const appName = process.env.APP_NAME!;
    const {
      data: { data: vaultSecrets },
    }: any = await Vault.read(`kv/data/erp/${appName}/${environment}`);
    const currentEnv = env[appName];
    switch (appName) {
      case 'ms-0':
        currentEnv.auth.jwt.publicKey = currentEnv.auth.jwt.publicKey
          ? currentEnv.auth.jwt.publicKey
          : vaultSecrets.publicKey;
        currentEnv.session.redis.url = currentEnv.session.redis.url
          ? currentEnv.session.redis.url
          : vaultSecrets.redis_url;
        currentEnv.session.cookie_name = currentEnv.session.cookie_name
          ? currentEnv.session.cookie_name
          : vaultSecrets.cookie_name;
        currentEnv.session.cookie_secret = currentEnv.session.cookie_secret
          ? currentEnv.session.cookie_secret
          : vaultSecrets.cookie_secret;
        break;
      case 'ms-1':
        currentEnv.databases.mongo.uri = currentEnv.databases.mongo.uri
          ? currentEnv.databases.mongo.uri
          : vaultSecrets.mongo_uri;
        currentEnv.auth.jwt.privateKey = currentEnv.auth.jwt.privateKey
          ? currentEnv.auth.jwt.privateKey
          : vaultSecrets.privateKey;
        currentEnv.auth.jwt.publicKey = currentEnv.auth.jwt.publicKey
          ? currentEnv.auth.jwt.publicKey
          : vaultSecrets.publicKey;
        break;
      case 'ms-3':
        currentEnv.databases.mongo.uri = currentEnv.databases.mongo.uri
          ? currentEnv.databases.mongo.uri
          : vaultSecrets.mongo_uri;
        currentEnv.auth.jwt.publicKey = currentEnv.auth.jwt.publicKey
          ? currentEnv.auth.jwt.publicKey
          : vaultSecrets.publicKey;
        break;
      case 'ms-4':
        currentEnv.databases.mongo.uri = currentEnv.databases.mongo.uri
          ? currentEnv.databases.mongo.uri
          : vaultSecrets.mongo_uri;
        currentEnv.auth.jwt.publicKey = currentEnv.auth.jwt.publicKey
          ? currentEnv.auth.jwt.publicKey
          : vaultSecrets.publicKey;
        break;
    }
  } catch (error) {
    sendError(TypeErrorMessage.VAULT);
  }
};

const readEnvFile = async () => {
  return new Promise<void>((resolve, reject) => {
    const env = dotenv.config();
    if (env.error) {
      reject(env.error);
    }
    resolve();
  });
};

File: packages/common/erp-common-ms-utils-py/gmahechas_erp_common_ms_utils_py/helpers/env.py
from os import environ
from importlib import import_module
from .vault import Vault, init_vault


class Env:
    _instance = None

    def __init__(self, config):
        if Env._instance is not None:
            raise Exception("Env already initialized")
        else:
            Env._instance = config

    def enrich(self, app_name, vault_secrets):
        current_env = self._instance[app_name]
        if app_name == "ms-3":
            current_env["databases"]["mongo"]["uri"] = current_env["databases"]["mongo"]["uri"] if current_env["databases"]["mongo"]["uri"] is not None else vault_secrets["mongo_uri"]
            current_env["auth"]["jwt"]["publicKey"] = current_env["auth"]["jwt"]["publicKey"] if current_env["auth"]["jwt"]["publicKey"] is not None else vault_secrets["publicKey"]

    @staticmethod
    def get():
        if Env._instance is None:
            raise Exception("Env not initialized")
        else:
            return Env._instance


def init_env():
    flask_app, environment, app_name = environ["FLASK_APP"], environ["ENVIRONMENT"], environ["APP_NAME"]
    module = import_module(f"{flask_app}.environments.{environment}", ".")
    env = Env(module.config)

    init_vault()
    vault_secrets = Vault.read(f"erp/{app_name}/{environment}")
    env.enrich(app_name, vault_secrets)
